design/authenticators/storing_non_sensitive_info.md

```diff
@@ -0,0 +1,331 @@
+
+- [Store Authenticators non-sensitive information](#store-authenticators-non-sensitive-information)
+  * [Introduction](#introduction)
+  * [Solution](#solution)
+    + [Authenticator modification](#authenticator-modification)
+    + [Backwards compatibility](#backwards-compatibility)
+  * [Test plan](#test-plan)
+  * [Performance](#performance)
+    + [Configuration](#configuration)
+    + [Authentication](#authentication)
+  * [Security](#security)
+  * [Documentation](#documentation)
+  * [Version Update](#version-update)
+  * [Delivery Plan](#delivery-plan)
+  
+## Introduction
+
+In several authenticators we need to store some information for their
+authentication process. For example, in `authn-oidc` we have 2 variables
+that store this kind of data:
+  - provider-uri
+  - id-token-user-property
+  
+These 2 variables are defined in the authenticator policy and are loaded with
+values before users can authenticate with the authenticator. However, this data
+is not sensitive and was stored in variables just because we didn't think of a 
+better solution. As the data is not sensitive, it makes sense to store them in
+other methods than in variables, which can improve the UX of the authenticator
+configuration and make another simplicity step.
+
+## Solution
+
+The suggested solution is to store this data in the authenticator webservice's
+annotations. 
+
+For example, let's look at a policy that defines an `authn-oidc`
+authenticator:
+
+```
+- !policy
+  id: conjur/authn-oidc/keycloak
+  body:
+  - !webservice
+    annotations:
+      description: Authentication service for Keycloak, based on Open ID Connect.
+
+  - !variable
+    id: provider-uri
+
+  - !variable
+    id: id-token-user-property
+
+  - !group
+    id: users
+    annotations:
+      description: Group of users who can authenticate using the authn-oidc/keycloak authenticator
+
+  - !permit
+    role: !group users
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+After loading the policy above, the operator will load the variables with values:
+```
+conjur variable values add conjur/authn-oidc/keycloak/provider-uri "some-provider-uri"
+conjur variable values add conjur/authn-oidc/keycloak/id-token-user-property "some-property"
+```
+
+In the suggested approach, the policy will not have variables, and will be defined
+as follows:
+```
+- !policy
+  id: conjur/authn-oidc/keycloak
+  body:
+  - !webservice
+    annotations:
+      description: Authentication service for Keycloak, based on Open ID Connect.
+      provider-uri: some-provider-uri
+      id-token-user-property: some-property
+
+  - !group
+    id: users
+    annotations:
+      description: Group of users who can authenticate using the authn-oidc/keycloak authenticator
+
+  - !permit
+    role: !group users
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+This improves the UX of the authenticator configuration for the following reasons:
```

Comment:
@orenbm I wonder if this suggestion solves the right problem. The first UX concern is related to needing a few steps for updating multiple variables. And the second UX concern can be done by bulk fetching all the variables, which Conjur supports today, just not in through the CLI.

If we would add bulk update of variables to Conjur, and added bulk update and retrieval to the Conjur CLI, wouldn't that solve the two authenticator UX concerns that were raised here? We could improve the UX even further with the suggestions I mentioned above.

Reply:
how is that a better UX? In your suggestion the user still needs to run a command to load the variables so it's still one more step (and a tedious one) in the configuration.

Furthermore, with the webservice annotation we get a visible and readable key-value map of metadata. in your suggestion we don't get that.

I see how your suggestion is better than our current process but not how it's better than the one suggested in this doc (and also - your suggestion will require more dev time).

Reply:
I think that my suggestion provides a better UX because it abstracts away from the user, how the authenticator is built and how it is saves in Conjur. The user would not need to understand the policy structure of the authenticators and would not need to understand if a value needs to go into an annotation or a variable. Instead the user would only need to be familiar with a single command that would ask from him/her all the required parameters. So it's a single command that would define and config the authenticator accordingly.

If we want a way to see variable values more easily, then maybe what we should really do, is add an API that gets the policy yaml and if a certain parameter is specified (and if you have permissions to read the variables) an API that would return both the policy and the variable values in it. We don't have to move data from one place to another, just because we like better how it looks, we can improve the UX of our existing tools and make them more convenient and friendly